# Day 12 Assignment - Hitik Panchal

In [23]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from scipy.stats import mannwhitneyu 
from scipy.stats import ttest_ind

# Reading the Data

In [7]:
gen_data=pd.read_csv('general_data.csv') 
gen_data.head() 

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [8]:
gen_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

# Cleaning the Data

In [9]:
gen_data.isnull().any()

Age                        False
Attrition                  False
BusinessTravel             False
Department                 False
DistanceFromHome           False
Education                  False
EducationField             False
EmployeeCount              False
EmployeeID                 False
Gender                     False
JobLevel                   False
JobRole                    False
MaritalStatus              False
MonthlyIncome              False
NumCompaniesWorked          True
Over18                     False
PercentSalaryHike          False
StandardHours              False
StockOptionLevel           False
TotalWorkingYears           True
TrainingTimesLastYear      False
YearsAtCompany             False
YearsSinceLastPromotion    False
YearsWithCurrManager       False
dtype: bool

In [10]:
gen_data.fillna(0 , inplace=True)

In [11]:
gen_data.isnull().any()

Age                        False
Attrition                  False
BusinessTravel             False
Department                 False
DistanceFromHome           False
Education                  False
EducationField             False
EmployeeCount              False
EmployeeID                 False
Gender                     False
JobLevel                   False
JobRole                    False
MaritalStatus              False
MonthlyIncome              False
NumCompaniesWorked         False
Over18                     False
PercentSalaryHike          False
StandardHours              False
StockOptionLevel           False
TotalWorkingYears          False
TrainingTimesLastYear      False
YearsAtCompany             False
YearsSinceLastPromotion    False
YearsWithCurrManager       False
dtype: bool

In [12]:
gen_data.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4405    False
4406    False
4407    False
4408    False
4409    False
Length: 4410, dtype: bool

In [13]:
gen_data.drop_duplicates()

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               0.0                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

In [20]:
yes_data = pd.DataFrame(gen_data[gen_data['Attrition']=='Yes'])
yes_data

Age Attrition     BusinessTravel              Department  \
1      31       Yes  Travel_Frequently  Research & Development   
6      28       Yes      Travel_Rarely  Research & Development   
13     47       Yes         Non-Travel  Research & Development   
28     44       Yes  Travel_Frequently  Research & Development   
30     26       Yes      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4381   29       Yes      Travel_Rarely  Research & Development   
4386   33       Yes      Travel_Rarely                   Sales   
4388   33       Yes      Travel_Rarely                   Sales   
4391   32       Yes      Travel_Rarely                   Sales   
4402   37       Yes  Travel_Frequently                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
1                   10          1  Life Sciences              1           2   
6                   11          2        Medical              1           7   
13                   1          1        Medical              1          14   
28                   1          2        Medical              1          29   
30                   4          3        Medical              1          31   
...                ...        ...            ...            ...         ...   
4381                 7          1  Life Sciences              1        4382   
4386                11          4      Marketing              1        4387   
4388                 1          3  Life Sciences              1        4389   
4391                23          1  Life Sciences              1        4392   
4402                 2          3      Marketing              1        4403   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
1     Female  ...                 0.0      Y                23              8   
6       Male  ...                 2.0      Y                20              8   
13      Male  ...                 1.0      Y                11              8   
28      Male  ...                 3.0      Y                14              8   
30      Male  ...                 2.0      Y                11              8   
...      ...  ...                 ...    ...               ...            ...   
4381  Female  ...                 1.0      Y                21              8   
4386    Male  ...                 5.0      Y                21              8   
4388    Male  ...                 7.0      Y                11              8   
4391    Male  ...                 0.0      Y                11              8   
4402    Male  ...                 6.0      Y                11              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
1                    1               6.0                      3   
6                    1               5.0                      2   
13                   2              10.0                      4   
28                   1              19.0                      2   
30                   0               5.0                      5   
...                ...               ...                    ...   
4381                 0               4.0                      2   
4386                 0               8.0                      2   
4388                 0              13.0                      2   
4391                 0               4.0                      2   
4402                 1              17.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
1                  5                        1                     4  
6                  0                        0                     0  
13                10                        9                     9  
28                 1                        0                     0  
30                 3                        0                     2  
...              ...                      ...                   ..

In [22]:
no_data = pd.DataFrame(gen_data[gen_data['Attrition']=='No']) 
no_data

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
5      46        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
5                    8          3  Life Sciences              1           6   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
5     Female  ...                 3.0      Y                13              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
5                    0              28.0                      5   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               0.0                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
5                  7                        7                     7  
...              ...                      ...                   ..

# Statistical Tests


## Seperate T-Test

__Test 1 :__ Attrition vs Distance From Home 

In [33]:
a1 = yes_data['DistanceFromHome']
a2 = no_data['DistanceFromHome']

tval, p = ttest_ind(a2,a1)

print("The T value is : %.3f" % tval)
print("The p value is : %.3f" % p)

The T value is : 0.646
The p value is : 0.518


__As the P value is 0.518 , which is > than 0.05, the Ha is rejected and H0 is accepted__

H0: There is no significant differences in the Distance From Home between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Distance From Home between Attrition (Y) and Attirition (N)

__Test 2 :__ Attrition vs Monthly Income

In [34]:
a1 = yes_data['MonthlyIncome']
a2 = no_data['MonthlyIncome']

tval, p = ttest_ind(a2,a1)

print("The T value is : %.3f" % tval)
print("The p value is : %.3f" % p)

The T value is : 2.071
The p value is : 0.038


__As the P value is 0.038 , which is < than 0.05, the Ha is accepted and H0 is rejected__

H0: There is no significant differences in the Monthly Income between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Monthly Income between Attrition (Y) and Attirition (N)

__Test 3 :__ Attrition vs Years at Company

In [35]:
a1 = yes_data['YearsAtCompany']
a2 = no_data['YearsAtCompany']

tval, p = ttest_ind(a2,a1)

print("The T value is : %.3f" % tval)
print("The p value is : %.3f" % p)

The T value is : 9.004
The p value is : 0.000


__As the P value is 0.0 , which is < than 0.05, the Ha is accepted and H0 is rejected__

H0: There is no significant differences in the Years at Company between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Years at Company between Attrition (Y) and Attirition (N)

__Test 4 :__ Attrition vs Years with Current Manager 

In [32]:
a1 = yes_data['YearsWithCurrManager']
a2 = no_data['YearsWithCurrManager']

tval, p = ttest_ind(a2,a1)

print("The T value is : %.3f" % tval)
print("The p value is : %.3f" % p)

The T value is : 10.499
The p value is : 0.000


__As the P value is 0.0 , which is < than 0.05, the Ha is accepted and H0 is rejected__

H0: There is no significant differences in the Years with Current Manager between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Years with Current Manager between Attrition (Y) and Attirition (N)

## Mann-Whitney Test

__Test 1 :__ Attrition vs Distance From Home 

In [37]:
a1 = yes_data['DistanceFromHome']
a2 = no_data['DistanceFromHome']

uval , p = mannwhitneyu(a1,a2)

print("The U value is : %.3f" % uval)
print("The p value is : %.3f" % p)

The U value is : 1312110.000
The p value is : 0.463


__As the P value is 0.463 , which is > than 0.05, the Ha is rejected and H0 is accepted__

H0: There is no significant differences in the Distance From Home between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Distance From Home between Attrition (Y) and Attirition (N)

__Test 2 :__ Attrition vs Monthly Income

In [38]:
a1 = yes_data['MonthlyIncome']
a2 = no_data['MonthlyIncome']

uval , p = mannwhitneyu(a1,a2)

print("The U value is : %.3f" % uval)
print("The p value is : %.3f" % p)

The U value is : 1264900.500
The p value is : 0.054


__As the P value is 0.054 , which is > than 0.05, the Ha is rejected and H0 is accepted__

H0: There is no significant differences in the Monthly Income between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Monthly Income between Attrition (Y) and Attirition (N)

__Test 3 :__ Attrition vs Years at Company

In [39]:
a1 = yes_data['YearsAtCompany']
a2 = no_data['YearsAtCompany']

uval , p = mannwhitneyu(a1,a2)

print("The U value is : %.3f" % uval)
print("The p value is : %.3f" % p)

The U value is : 923238.000
The p value is : 0.000


__As the P value is 0.0 , which is < than 0.05, the Ha is accepted and H0 is rejected__

H0: There is no significant differences in the Years at Company between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Years at Company between Attrition (Y) and Attirition (N)

__Test 4 :__ Attrition vs Years with Current Manager 

In [40]:
a1 = yes_data['YearsWithCurrManager']
a2 = no_data['YearsWithCurrManager']

uval , p = mannwhitneyu(a1,a2)

print("The U value is : %.3f" % uval)
print("The p value is : %.3f" % p)

The U value is : 957253.500
The p value is : 0.000


__As the P value is 0.0 , which is < than 0.05, the Ha is accepted and H0 is rejected__

H0: There is no significant differences in the Years with Current Manager between Attrition (Y) and Attirition (N)

Ha: There is significant differences in the Years with Current Manager between Attrition (Y) and Attirition (N)